File: java/src/com/algos/topics/li09_linkedlist/revised/L143_ReorderList.java
package  com.algos.topics.li09_linkedlist.revised;

import com.algos.topics.models.ListNode;

public class L143_ReorderList {

	public static void main(String[] args) {
		L143_ReorderList obj = new L143_ReorderList();
		SingleLinkedList list = new SingleLinkedList();
		list.insertAfterEnd(1);
		list.insertAfterEnd(2);
		list.insertAfterEnd(3);
		list.insertAfterEnd(4);

		list.print();
		obj.reorderList(list.head);
		System.out.println();
		list.print();
	}

	// Problem: https://leetcode.com/problems/reorder-list/
	// Idea: Split the list, reverse the second list and merge.
	public void reorderList(ListNode head1) {
        if(head1.next == null || head1.next.next == null)
            return;

        ListNode head2 = split(head1);
        head2 = reverse(head2);

        ListNode curr1 = head1, curr2 = head2;
        while(curr1 != null && curr2 != null){
            ListNode next1 = curr1.next;
            ListNode next2 = curr2.next;
            curr1.next = curr2;
            curr2.next = next1;

            curr1 = next1;
            curr2 = next2;
        }
    }

    private ListNode reverse(ListNode head2) {
        ListNode prev = null, curr = head2;
        while(curr != null){
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }

    private ListNode split(ListNode head1) {
        ListNode slow = head1;
        ListNode fast = head1.next;         // delay slow by one step
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode head2 = slow.next;
        slow.next = null;
        return head2;
    }

}

File: dsa/concepts/09_linkedlist/python/l143_reorder_list.py
"""
Problem: https://leetcode.com/problems/reorder-list/

Idea:
Split the list, reverse the second list and merge.
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def reorder_list(head: ListNode) -> None:
    if not head or not head.next or not head.next.next:
        return

    # Step 1: split the list
    head2 = split(head)

    # Step 2: reverse second half
    head2 = reverse(head2)

    # Step 3: merge alternately
    curr1, curr2 = head, head2
    while curr1 and curr2:
        next1 = curr1.next
        next2 = curr2.next

        curr1.next = curr2
        curr2.next = next1

        curr1 = next1
        curr2 = next2


def reverse(head: ListNode) -> ListNode:
    prev = None
    curr = head
    while curr:
        nxt = curr.next
        curr.next = prev
        prev = curr
        curr = nxt
    return prev


def split(head: ListNode) -> ListNode:
    slow = head
    fast = head.next      # delay slow by one step
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next

    head2 = slow.next
    slow.next = None
    return head2


# Example usage
if __name__ == "__main__":
    # 1 -> 2 -> 3 -> 4
    n1 = ListNode(1)
    n2 = ListNode(2)
    n3 = ListNode(3)
    n4 = ListNode(4)

    n1.next = n2
    n2.next = n3
    n3.next = n4

    reorder_list(n1)

    curr = n1
    while curr:
        print(curr.val, end=" ")
        curr = curr.next
    # Output: 1 4 2 3
